src/web_socket.toit

```diff
@@ -0,0 +1,304 @@
+// Copyright (C) 2022 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the LICENSE file.
+
+import binary show BIG_ENDIAN
+import net.tcp
+import reader
+import writer
+
+OPCODE_CONTINUATION_ ::= 0
+OPCODE_TEXT_         ::= 1
+OPCODE_BINARY_       ::= 2
+OPCODE_CLOSE_        ::= 8
+OPCODE_PING_         ::= 9
+OPCODE_PONG_         ::= 10
+FIN_FLAG_            ::= 0x80
+EIGHT_BYTE_SIZE_     ::= 127
+TWO_BYTE_SIZE_       ::= 126
+MAX_ONE_BYTE_SIZE_   ::= 125
+
+/**
+A WebSocket connection.
+A bidirectional socket connection capable of sending binary or text messages
+  according to RFC 6455.
+Obtained from the $Client.web_socket method.
+Currently the server side setup is not implemented.
+Currently does not implement ping and pong packets.
+*/
+class WebSocket:
+  socket_ /tcp.Socket
+  pending_ /ByteArray? := null
+  current_writer_ /WebSocketWriter? := null
+  current_reader_ /WebSocketReader? := null
+
+  constructor .socket_:
+
+  read_ -> ByteArray?:
+    if pending_:
+      result := pending_
+      pending_ = null
+      return result
+    return socket_.read
+
+  unread_ byte_array/ByteArray -> none:
+    assert: pending_ == null
+    pending_ = byte_array
+
+  /**
+  Returns a reader for the next message sent to us on the WebSocket.
+  Returns null if the connection is closed.
+  Should not be called until the previous reader has been fully read.
+  */
+  receive -> WebSocketReader?:
+    if current_reader_ != null: throw "PREVIOUS_READER_NOT_FINISHED"
+    fragment_reader := receive_
+    if fragment_reader == null: return null
+    if fragment_reader.is_ping or fragment_reader.is_pong:
+      throw "UNIMPLEMENTED_PING"
```

Comment:
Not saying this is important, but I would expect that implementing 'pong' should be trivial.